bindings/src/main/java/io/pravega/storage/filesystem/FileSystemStorage.java

```diff
@@ -458,5 +466,225 @@ private Void doDelete(SegmentHandle handle) throws IOException {
+            return withRecovery(streamSegmentName, super::doGetStreamSegmentInfo);
+        }
+
+        @Override
+        protected Void doConcat(SegmentHandle targetHandle, long offset, String sourceSegment) throws IOException {
+            return withRecovery(sourceSegment, source -> super.doConcat(targetHandle, offset, source));
+        }
+
+        @Override
+        protected boolean doExists(String streamSegmentName) {
+            return super.doExists(streamSegmentName) || super.doExists(getTempSegmentName(streamSegmentName));
+        }
+
+        @Override
+        public boolean supportsReplace() {
+            return true;
+        }
+
+        @Override
+        public void replace(@NonNull SegmentHandle segment, @NonNull BufferView contents) throws StreamSegmentException {
+            String segmentName = segment.getSegmentName();
+            execute(segment.getSegmentName(), () -> replaceExistingFile(segmentName, contents));
+        }
+
+        @Override
+        public SyncStorage withReplaceSupport() {
+            return this;
+        }
+
+        //endregion
+
+        private <T, TEx extends Exception> T withRecovery(String segmentName, RecoverableAction<T, TEx> toExecute) throws TEx {
+            try {
+                return toExecute.apply(segmentName);
+            } catch (Exception ex) {
+                String tmpName = getTempSegmentName(segmentName);
+                if (isFileNotFoundException(ex) && super.doExists(tmpName)) {
+                    // Incomplete replace operation detected. Finish it up.
+                    log.info("Incomplete replace operation detected for '{}'. Finalizing.", segmentName);
+                    finalizeRename(tmpName, segmentName);
+
+                    // Should be done now. Retry original operation.
+                    log.debug("Replace finalized for '{}'. Retrying operation.", segmentName);
+                    return toExecute.apply(segmentName);
+                } else {
+                    throw ex;
+                }
+            }
+        }
+
+        private Void replaceExistingFile(String segmentName, BufferView contents) throws IOException, StreamSegmentException {
+            boolean baseExists = super.doExists(segmentName);
+            boolean shouldReseal = baseExists && super.getStreamSegmentInfo(segmentName).isSealed();
+
+            // Check temp file already exists.
+            String tmpSegmentName = getTempSegmentName(segmentName);
+            if (super.doExists(tmpSegmentName)) {
+                if (baseExists) {
+                    // We have both a temp file and a segment file. This is most likely the result of an incomplete replace,
+                    // however we still have the original segment file around. It is safe to delete the temp file now.
+                    log.debug("Incomplete replace operation detected for '{}'. Deleting temp file before new replace attempt.", segmentName);
```

Comment:
Since this case is not the regular, expected path, I'd say this log message is at least `info`.

Reply:
Done.